tests/unit/apps/test_base.py

```diff
@@ -485,3 +485,127 @@ def test_check_mismatched_versions(mock_os_release_units, model):
     )
 
     assert app._check_mismatched_versions([units["my-app/0"]]) is None
+
+
+@patch("cou.apps.base.OpenStackApplication._verify_channel", return_value=None)
+def test_get_charm_hub_migration_ch(_, model):
+    """Application is already from charm hub."""
+    app_name = "app"
+    app = OpenStackApplication(
+        app_name, "", app_name, "ussuri/stable", {}, {}, model, "ch", "focal", [], {}, "1"
+    )
+    assert app._get_charm_hub_migration() == PreUpgradeStep()
+
+
+@patch("cou.apps.base.OpenStackApplication._verify_channel", return_value=None)
+@patch("cou.apps.base.OpenStackApplication.current_os_release", new_callable=PropertyMock)
+def test_get_charm_hub_migration_cs(current_os_release, _, model):
+    """Application is already from charm store."""
+    current_os_release.return_value = OpenStackRelease("ussuri")
+    app_name = "app"
+    app = OpenStackApplication(
+        app_name, "", app_name, "ussuri/stable", {}, {}, model, "cs", "focal", [], {}, "1"
+    )
+    assert app._get_charm_hub_migration() == PreUpgradeStep(
+        f"Migrate '{app.name}' from charmstore to charmhub",
+        coro=model.upgrade_charm(app.name, app.possible_current_channel, switch=f"ch:{app.charm}"),
+    )
+
+
+@patch("cou.apps.base.OpenStackApplication._verify_channel", return_value=None)
+@patch("cou.apps.base.OpenStackApplication.current_os_release", new_callable=PropertyMock)
+def test_get_change_to_openstack_channels_latest_stable(current_os_release, _, model):
+    """Application using latest/stable channel."""
```

Comment:
For clarity, you could explain what's the expectation is, e.g. "Applications using latest/stable should be switched to a release-specific channel"